spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java

```diff
@@ -170,4 +174,34 @@ public static boolean useTimestampWithoutZoneInNewTables(RuntimeConfig sessionCo
     return false;
   }
 
+  /**
+   * Pulls any Catalog specific overrides for the Hadoop conf from the current SparkSession, which can be
+   * set via spark.sql.catalog.$catalogName.hadoop.*
+   *
+   * The SparkCatalog allows for hadoop configurations to be overridden per catalog, by setting
+   * them on the SQLConf, where the following will add the property "fs.default.name" with value
+   * "hdfs://hanksnamenode:8020" to the catalog's hadoop configuration.
+   *   SparkSession.builder()
+   *     .config(s"spark.sql.catalog.$catalogName.hadoop.fs.default.name", "hdfs://hanksnamenode:8020")
+   *     .getOrCreate()
+   * @param spark The current Spark session
+   * @param catalogName Name of the catalog to find overrides for.
+   * @return the Hadoop Configuration that should be used for this catalog, with catalog specific overrides applied.
+   */
+  public static Configuration hadoopConfCatalogOverrides(SparkSession spark, String catalogName) {
+    // Find keys for the catalog intended to be hadoop configurations
+    final String hadoopConfCatalogPrefix = String.format("%s.%s.%s", SPARK_CATALOG_CONF_PREFIX, catalogName, "hadoop.");
+    final Configuration conf = spark.sessionState().newHadoopConf();
+    // settings is a java.util.Collections.synchronizedMap and needs to be wrapped in `synchronized`.
+    synchronized (spark.sqlContext().conf().settings()) {
```

Comment:
I added `synchronized` as mentioned by @szehon-ho, but looking at the overridden definition of `forEach` for `java.util.Collections.SynchronizedMap`. I don't think this is necessary.

Here's the definition of forEach copied from my IDE. Note that it already synchronizes on itself (which is the `settings()` result here).

```java
        @Override
        public void forEach(BiConsumer<? super K, ? super V> action) {
            synchronized (mutex) {m.forEach(action);}
        }
```

Reply:
Yea looks good to me to remove it, thanks for looking into it.